opacus/grad_sample/grad_sample_module.py

```diff
@@ -260,6 +298,94 @@ def is_supported(cls, module: nn.Module) -> bool:
         """Check if this module is supported"""
         return type(module) in cls.GRAD_SAMPLERS or type(module) is DPLSTM
 
+    def ddp_backward_callback(self, engine, p, grad):
+        """
+        This hook operates like PrivacyEngine.step(), but on a single layer:
+        1. clip_and_accumulate
+        2. get the clip_values with clipper.pre_step()
+        3. add the noise
+        """
+
+        # Get the norm of the gradient for each sample for this layer
+        all_norms = calc_sample_norms(
+            named_params=((None, p.grad_sample),),
+            flat=False,
+        )
+
+        # Get the constant clipping factor for a single layer
+        clipping_factor = engine.clipper.norm_clipper.calc_clipping_factors(all_norms)
+        assert len(clipping_factor) == 1
+        clip_factor = clipping_factor[0]
+        batch_size = p.grad_sample.shape[0]
+
+        # Do the clipping
+        summed_grad = engine.clipper._weighted_sum(clip_factor, p.grad_sample)
+
+        # accumulate the summed gradient for this mini-batch
+        if hasattr(p, "summed_grad"):
+            p.summed_grad += summed_grad
+        else:
+            p.summed_grad = summed_grad
+
+        del p.grad_sample
+
+        # Average (or sum) across the batch
+        # TODO: add support for Poisson batch sampling
+        res = engine.clipper._scale_summed_grad(p.summed_grad, batch_size)
+
+        del p.summed_grad
+
+        # NOTE: `self.clipper.pre_step()` returns a tensor with n=n_params clip_values
+        # where each clip_value is the L2 norm of `threshs`:
+        #   `max_norm = threshs.new_full((n,), threshs.norm(2))`
+        # With per layer clipping, `threshs` is a tensor with n flat_values
+        assert len(engine.clipper.norm_clipper.thresholds) == 1
+        clipping_thresh = engine.clipper.norm_clipper.thresholds[0]
+        clip_value = (self.n_params ** 0.5) * clipping_thresh
```

Comment:
I agree that this snippet is a bit confusing. However, it seems that this behavior (increasing the multiplier by `sqrt(n_params)`) is the current behavior when we use `_Dynamic_Clipper_` with `clip_per_layer=True` and `max_grad_norm=C` for some float C. 
What do you think @thomasflynn918? 

In any case, I will try to make this part clearer when I refactor the hook -- especially if we have a different threshold for each layer.

Reply:
I think you're indeed right, that `_Dynamic_Clipper_` does the same. But to me that's more of an argument to change `_Dynamic_Clipper_` too - for the reasons I've outlined above I think it's more consistent if by `max_grad_norm` we always mean "L2 norm of the entire grad vector" (unless it's a list, then it's reasonable to expect different interpretation)

Reply:
Good point. Since I want to support per-layer clipping in the general case, I am now expecting `max_grad_norm` to be a list. I dropped support for `_Dynamic_Clipper_` because it was only making sense in the cases where it was equal to `ConstantPerLayerClipper` (we can still implement distributed versions of the dynamic schemes later).